### ISSUE TITLE
[LLM] Support LLM routing

### ISSUE THREAD

**Issue by xingyaoww** [All-Hands-AI/OpenHands#4184]
**What problem or use case are you trying to solve?**

Not Diamond intelligently identifies which LLM is best-suited to respond to any given query. We want to implement a mechanism in OpenHands to support this type of "LLM" selector.

**Describe the UX of the solution you'd like**

Ideally, use should define a "LLMRouter" as a special type of LLM with some special configs (e.g., multiple keys for different providers). And user can just put in keys, and select that router, and OpenHands will automatically use that going forward.

**Do you have thoughts on the technical implementation?**

Modify https://github.com/All-Hands-AI/OpenHands/blob/main/openhands/llm/llm.py, as well as config related files under https://github.com/All-Hands-AI/OpenHands/tree/main/openhands/core/config.

You should probably use `model_select` (from notdiamond API) rather than `create` to be compatible with existing LiteLLM calls.

**Describe alternatives you've considered**

**Additional context**

Here's the documentation from NotDiamond

```
# -*- coding: utf-8 -*-
"""Getting started with Not Diamond

Automatically generated by Colab.

Original file is located at
    https://colab.research.google.com/drive/1Ao-YhYF_S6QP5UGp_kYhgKps_Sw3a2RO

# **Setting up**
"""

!pip install -q notdiamond[create] --upgrade

import os

os.environ["NOTDIAMOND_API_KEY"] = 'YOUR_NOTDIAMOND_API_KEY'
os.environ["OPENAI_API_KEY"] = 'YOUR_OPENAI_API_KEY'
os.environ["ANTHROPIC_API_KEY"] = 'YOUR_ANTHROPIC_API_KEY'
os.environ["PPLX_API_KEY"] = 'YOUR_PERPLEXITY_API_KEY'

"""# **Automatic routing and model calling**

We'll start by defining the routing `NotDiamond` client, which will function like a 'meta-LLM' that ensembles together the best of multiple models.

We then add specific LLM targets for routing. Not Diamond works with any LLM in [this list](https://notdiamond.readme.io/docs/llm-models).
"""

from notdiamond import NotDiamond

client = NotDiamond()

llm_providers = [
    'openai/gpt-4o',
    'anthropic/claude-3-5-sonnet-20240620',
    'openai/gpt-4o-mini',
    'perplexity/llama-3.1-sonar-large-128k-online'
]

"""Next, let's call the client by passing in an array of messages and our target models:"""

result, session_id, provider = client.chat.completions.create(
    messages=[
        {"role": "system", "content": "You are a helpful assistant."},
        {"role": "user", "content": "Concisely explain merge sort."}  # Adjust as desired
    ],
  	model=llm_providers
)

print("LLM called: \n", provider.model)
print("\nLLM output: \n", result.content)

"""If we pass in a different question, `NotDiamond` will recommend a different model:"""

result, session_id, provider = client.chat.completions.create(
    messages=[
        {"role": "system", "content": "You are a helpful assistant."},
        {"role": "user", "content": "What's the weather like in SF today?"}
    ],
  	model=llm_providers
)

print("LLM called: \n", provider.model)
print("\nLLM output: \n", result.content)

"""# **Defining tradeoffs**

We can also define tradeoffs for cost or latency to adjust routing to our preferences:
"""

result, session_id, provider = client.chat.completions.create(
    messages=[
        {"role": "system", "content": "You are a helpful assistant."},
        {"role": "user", "content": "Concisely explain merge sort."}
    ],
  	model=llm_providers,
  	tradeoff="cost" # Consider cheaper models when quality loss is negligible. Alternatively, you can use "latency".
)

print("LLM called: \n", provider.model)
print("\nLLM output: \n", result.content)

"""# **Routing recommendations with `model_select`**

Finally, we can also use `model_select` to return a recommended LLM for your prompt. You can then invoke that LLM using your own custom logic.
"""

session_id, provider = client.chat.completions.model_select(
    messages=[
        {"role": "system", "content": "You are a helpful assistant."},
        {"role": "user", "content": "Concisely explain merge sort."}
    ],
  	model=llm_providers
)

## Add more application logic here. e.g.:

print(f"LLM called: \n{provider.model}\n", )

match provider.model:
    case "claude-3-5-sonnet-20240620":
        print("Running custom_sonnet_3-5_invoke()...")
        # custom_sonnet_35_invoke()
    case 'gpt-4o':
        print("Running custom_gpt_4o_invoke()...")
        # custom_gpt_4o_invoke()
    case 'llama-3-sonar-large-32k-online':
        print("Running custom_pplx_llama3_invoke()...")
        # custom_pplx_llama3_invoke()
    case 'gpt-4o-mini':
        print("Running custom_gpt_4o_mini_invoke()...")
        # custom_gpt_4o_mini_invoke()
```

**Comment by tobitege** [comment]
@xingyaoww - also see #4109 where litellm's Router is being incorporated and also a config structure that could _maybe_ used here

**Comment by neubig** [comment]
Quick point of discussion: do we want to implement this within OpenHands? Or should we host a server with the router, like we host our proxy server for All Hands AI?

Personally I think the latter might be better. Doing this on the client side means that users have to acquire several different API keys and somehow configure them. This seems like a pain UI-wise, especially given that currently our configuration behavior is hard to understand: https://github.com/All-Hands-AI/OpenHands/issues/3220

**Comment by tobitege** [comment]
Hey @neubig 
this topic is similarly picked up here, but for the litellm Router (different purpose, but maybe similar router config). Not necessarily something for the UI, though.

**Comment by neubig** [comment]
Yeah, maybe NotDiamond could be implemented as a custom routing strategy within the LiteLLM proxy?

**Comment by xingyaoww** [comment]
yeah seems like a better approach (if we can get the cost propagation to work correctly). Close this for now then

**Comment by acompa** [comment]
Hi @xingyaoww @neubig, just caught this issue.

While our LLMConfigs [accept prices](https://github.com/Not-Diamond/notdiamond-python/blob/main/notdiamond/llms/config.py#L71-L72), they only help tune cost tradeoffs. You won't have to provide that parameter for public models - we track prices for [every model we support](https://docs.notdiamond.ai/docs/llm-models).

Beyond this, we're also happy to help you set up a routing integration with Not Diamond's API. Just let me know if that interests you. 

As for LiteLLM, we've actually been discussing an integration with them [since July](https://github.com/BerriAI/litellm/pull/4971)! While waiting on their feedback, we've also implemented a simple integration [in our Python client](https://github.com/Not-Diamond/notdiamond-python/tree/main/notdiamond/toolkit/litellm) which might help you.

**Comment by neubig** [comment]
Thanks @acompa , I do think we'd be interested in at least running an evaluation where we use NotDiamond as a backend and see if the results are better/cheaper than what we get now. If your API offers OpenAI compatible endpoints it should be pretty easy (we haven't looked super-carefully yet).

**Comment by acompa** [comment]
> Thanks @acompa , I do think we'd be interested in at least running an evaluation where we use NotDiamond as a backend and see if the results are better/cheaper than what we get now. If your API offers OpenAI compatible endpoints it should be pretty easy (we haven't looked super-carefully yet).

We do accept [OpenAI-style requests with `messages`](https://docs.notdiamond.ai/docs/quickstart#sending-your-first-not-diamond-api-request) at our `model_select` endpoint. We're not a proxy, though, so at the moment we only support [`create` via a Langchain integration](https://docs.notdiamond.ai/docs/model_select-vs-create).

**Comment by neubig** [comment]
Cool, thanks! I'll re-open this as I think that whatever way we implement it'd be interesting to see if model routing helps.

**Comment by acompa** [comment]
Excellent. As you begin your evaluation, note that we offer two approaches to AI model routing:

Our [out-of-the-box router](https://www.notdiamond.ai/features#SOTA-model-routing) has been trained on generalist, cross-domain data (including coding and non-coding tasks) to provide a strong "multi-model" multidisciplinary experience.

Secondly, OpenHands focuses on development applications, and so you might benefit from specialized routing trained on the distribution of your proprietary data. We also offer [custom routing](https://docs.notdiamond.ai/docs/router-training-quickstart) to serve these types of domain-targeted use cases as a higher-performance option beyond out-of-the-box routing.

We're happy to answer questions or support you in whichever of these approaches you evaluate.

**Comment by tobitege** [comment]
@neubig we could also look into the https://github.com/Not-Diamond/RoRF/ repo to start with (pair-wise routing) to start with?

**Comment by neubig** [comment]
I think the NotDiamond folks are working on this still.

**Comment by xingyaoww** [comment]
I think this is still in progress?

**Comment by neubig** [comment]
Yes, it is!

**Comment by enyst** [comment]
Is this WIP? Any updates would be great!

**Comment by xingyaoww** [comment]
@ryanhoangt can you share some progress here?

**Comment by ryanhoangt** [comment]
Yeah, Notdiamond team has helped us to train a custom router for a group of defined models, and that router works by choosing an LLM for a particular instance, and all the turns in the trajectory will be routed to that router. When I ran an eval it didn't work too well and the team said they'll help us debug but they're not available currently. So currently I'm working on an alternative approach to routing by interleaving calls to different LLMs in a trajectory using a fine-tuned LLM as the router. I tried to train a router for Sonnet 3.5 and Deepseek v3 and some preliminary empirical experiments show that it's possible to achieve the same performance while reducing the cost by 20-40%, but it's just on SWE-bench evaluation and we probably need more diverse data to be able to use it in production. I'm also trying to do more analysis and experiments on the result.

![Image](https://github.com/user-attachments/assets/ee4ba88f-6b2c-45c9-a70b-7d50a2bc380a)

**Comment by BradKML** [comment]
Are there ways to experiment with Wilmer or RouteLLM, and create a predictor to see when certain tasks are better off with smaller models? (ditto for thinking models)

**Comment by pcuci** [comment]
> I'm working on an alternative approach to perform routing by interleaving calls to different LLMs in a trajectory using a fine-tuned LLM as the router.

@ryanhoangt - how about autotuning as an alternative to pretraining routers? https://github.com/GenseeAI/cognify

I probably don't understand the design yet, but shouldn't the routing change during long trajectories?

**Comment by enyst** [comment]
To note here

> Quick point of discussion: do we want to implement this within OpenHands? Or should we host a server with the router, like we host our proxy server for All Hands AI?
> 
> Personally I think the latter might be better. Doing this on the client side means that users have to acquire several different API keys and somehow configure them. This seems like a pain UI-wise, especially given that currently our configuration behavior is hard to understand: [#3220](https://github.com/All-Hands-AI/OpenHands/issues/3220)

I see this as implemented within OpenHands. We have actually taken a couple of attempts to it in time, and IIRC a PR is in progress today as well.

By now, people have the ability to [configure multiple LLMs](https://github.com/All-Hands-AI/OpenHands/blob/848f69203366faa405cb535f125f35003c12379e/config.template.toml#L222) for a long time. The evaluation scripts always use any of the multiple LLMs configured on our machines.

The functionality is backend-only and command line-only. It's not exposed in the Web UI, and it's not read/considered by the `server` component, so it's not available to users when running UI.

It's in use with the CLI app and headless mode.

I'm not involved in the web frontend, and I don't know why we're not trying to expose it there. I do think configurable multiple LLMs and functionality built on it is very important functionality, and users without it are hitting all kinds of limitations.

**Comment by BradKML** [comment]
Cross-linking references cus there are task-based routing vs difficulty-based routing https://github.com/All-Hands-AI/OpenHands/issues/8524

**Comment by enyst** [comment]
@jpelletier1 Just to note, this is the same with PR 9738, we can remove 9738 from the roadmap

**Comment by jpelletier1** [comment]
@enyst thank you! We'll track #9738 on roadmap to eliminate redundancy.

**Comment by enyst** [comment]
That could work, though I would’ve suggested to track the issue, not one of the PRs for it. We’ve been making other attempts (PRs) at it in time, it’s not the easiest to crack. 😅

Although this time, I’m confident we’ll merge! 😂

**Comment by jpelletier1** [comment]
@enyst You are right - better to track this issue on roadmap, especially if there will be multiple PRs :) Updated.

**Comment by BradKML** [comment]
If routing is done, will issues from Qwen and DeepSeek be fixed as well? (different issue but a weird heads up) https://github.com/SomeOddCodeGuy/WilmerAI/discussions/26#discussioncomment-14041128

**Comment by BradKML** [comment]
Would "fallback" features be covered? https://github.com/All-Hands-AI/OpenHands/issues/4056
